runtime/cudaq/builder/kernel_builder.h

```diff
@@ -408,6 +408,12 @@ class kernel_builder : public details::kernel_builder_base {
   void NAME(QuakeValue &qubit) {                                               \
     std::vector<QuakeValue> empty;                                             \
     details::NAME(*opBuilder, empty, qubit, true);                             \
+  }                                                                            \
+  template <typename mod,                                                      \
+            typename =                                                         \
+                typename std::enable_if_t<std::is_same_v<mod, cudaq::adj>>>    \
+  void NAME(QuakeValue &&qubit) {                                              \
```

Comment:
I see this is copied from above, but would an explicit template specialization work just as well here?
```suggestion
  template <> void NAME<cudaq::adj>(QuakeValue &&qubit) {                                              \
```

Reply:
I was wondering if we could merge both into one with `void NAME(const QuakeValue& qubit);`

Reply:
@schweitzpgi It wouldn't work as it is since the surrounding class (`kernel_builder`) is also templated. 
I guess SFINAE as above is one of the cleanest ways to work around this according to this [SO post](https://stackoverflow.com/questions/5512910/explicit-specialization-of-template-class-member-function) :-)

@amccaskey Yes, we could do that. It'd require a change to the `CUDAQ_DETAILS_QIS_DECLARATION` and `CUDAQ_ONE_QUBIT_IMPL` to send on the `const` modifier (it is using this `QuakeValue` in a `const` manner already).

The changeset would look like this (from the HEAD of this branch)
```diff
--- a/runtime/cudaq/builder/kernel_builder.cpp
+++ b/runtime/cudaq/builder/kernel_builder.cpp
@@ -392,7 +392,7 @@ void applyOneQubitOp(ImplicitLocOpBuilder &builder, auto &&params, auto &&ctrls,
 
 #define CUDAQ_ONE_QUBIT_IMPL(NAME, QUAKENAME)                                  \
   void NAME(ImplicitLocOpBuilder &builder, std::vector<QuakeValue> &ctrls,     \
-            QuakeValue &target, bool adjoint) {                                \
+            const QuakeValue &target, bool adjoint) {                          \
     cudaq::info("kernel_builder apply {}", std::string(#NAME));                \
     auto value = target.getValue();                                            \
     auto type = value.getType();                                               \
diff --git a/runtime/cudaq/builder/kernel_builder.h b/runtime/cudaq/builder/kernel_builder.h
index 3088440..5407102 100644
--- a/runtime/cudaq/builder/kernel_builder.h
+++ b/runtime/cudaq/builder/kernel_builder.h
@@ -148,7 +148,7 @@ QuakeValue qalloc(ImplicitLocOpBuilder &builder, QuakeValue &size);
 
 #define CUDAQ_DETAILS_QIS_DECLARATION(NAME)                                    \
   void NAME(ImplicitLocOpBuilder &builder, std::vector<QuakeValue> &ctrls,     \
-            QuakeValue &target, bool adjoint = false);
+            const QuakeValue &target, bool adjoint = false);
 
 CUDAQ_DETAILS_QIS_DECLARATION(h)
 CUDAQ_DETAILS_QIS_DECLARATION(s)
@@ -405,15 +405,9 @@ public:
   template <typename mod,                                                      \
             typename =                                                         \
                 typename std::enable_if_t<std::is_same_v<mod, cudaq::adj>>>    \
-  void NAME(QuakeValue &qubit) {                                               \
+  void NAME(const QuakeValue &qubit) {                                         \
     std::vector<QuakeValue> empty;                                             \
     details::NAME(*opBuilder, empty, qubit, true);                             \
-  }                                                                            \
-  template <typename mod,                                                      \
-            typename =                                                         \
-                typename std::enable_if_t<std::is_same_v<mod, cudaq::adj>>>    \
-  void NAME(QuakeValue &&qubit) {                                              \
-    NAME<mod>(qubit);                                                          \
   }
```

Reply:
I think we should update and do it that way.

Reply:
Thanks for the clarification. Makes perfect sense.

Reply:
I've updated the code to use `const&` as suggested by Alex.